src/corelibs/U2Core/src/util/MsaRowUtils.cpp

```diff
@@ -327,6 +327,25 @@ bool MsaRowUtils::isGap(int dataLength, const U2MsaRowGapModel &gapModel, int po
     return false;
 }
 
+bool MsaRowUtils::isLeadingOrTrailingGap(int dataLength, const U2MsaRowGapModel &gapModel, int position) {
+    int gapsLength = 0;
+    foreach (const U2MsaGap &gap, gapModel) {
+        if (gap.offset == 0 && gap.offset <= position && position < gap.offset + gap.gap) {
```

Comment:
gap.offset <= position is not needed here.

Overall I would suggest not to copy-paste, but to write a simple/optimal/readable/documented code focused on the leading or trailing gap only. Here is an example:

CHECK(!gapModel.isEmpty(), false);
    if (gapModel[0].offset == 0 && position < gapModel[0].endPos()) {
        return true;    // leading gap.
    }
    int totalGapsLen = 0;
    for (const U2MsaGap& gap : gapModel) {
        totalGapsLen += gap.gap;
        if (position < gap.offset) {
            return false; // somethere in the middle.
        } 
    }
    return position >= dataLength + totalGapsLen; // trailing gap.